### PR TITLE
Add support for PTR query

### PR DESCRIPTION
**Pull request by mokeyish**
This PR allows us to do IP reverse lookup on the Hosts.

### CODE REVIEW

File: crates/resolver/src/hosts.rs [hunk 1]
```diff
@@ -47,14 +49,62 @@ impl Hosts {
...
-                let result = match query.query_type() {
-                    RecordType::A => val.a.clone(),
-                    RecordType::AAAA => val.aaaa.clone(),
-                    _ => None,
-                };
-
-                return result;
+            match query.query_type() {
+                RecordType::A | RecordType::AAAA => {
+                    if let Some(val) = self.by_name.get(query.name()) {
```
Comment (djc) [inline crates/resolver/src/hosts.rs:54]: Would like to take the opportunity to use `let-else` here to do an early return if `self.by_name.get()` yields `None`.

File: crates/resolver/src/hosts.rs [hunk 2]
```diff
@@ -47,14 +49,62 @@ impl Hosts {
...
+                            RecordType::A => val.a.clone(),
+                            RecordType::AAAA => val.aaaa.clone(),
+                            _ => None,
+                        };
+
+                        return result;
+                    }
+                }
+                RecordType::PTR => {
+                    if let Ok(ip) = query.name().parse_arpa_name() {
```
Comment (djc) [inline crates/resolver/src/hosts.rs:65]: Let's use `let-else` here to reduce rightward drift here, and also for the `hostnames.is_empty()` case.

File: crates/resolver/src/hosts.rs [hunk 3]
```diff
@@ -47,14 +49,62 @@ impl Hosts {
...
+                                        r.ip_addr().map(|it| it == ip).unwrap_or_default()
+                                    }),
+                                    None => false,
+                                },
+                            })
+                            .map(|(n, _)| n)
+                            .collect::<Vec<_>>();
+
+                        if !hostnames.is_empty() {
+                            hostnames.sort();
```
Comment (djc) [inline crates/resolver/src/hosts.rs:88]: Why is sorting a requirement here?

File: crates/resolver/src/hosts.rs [hunk 4]
```diff
@@ -47,14 +49,62 @@ impl Hosts {
...
+                            let records = hostnames
+                                .into_iter()
+                                .map(|hostname| {
+                                    Record::from_rdata(
+                                        query.name().clone(),
+                                        dns_lru::MAX_TTL,
+                                        RData::PTR(PTR(hostname.clone())),
+                                    )
+                                })
+                                .collect::<Vec<_>>();
```
Comment (djc) [inline crates/resolver/src/hosts.rs:98]: It looks like we can chain the `map()` here onto `hostnames`, avoiding the intermediate allocation.

**Review by djc — COMMENTED** [review]
I think this makes sense, would be nice clean up the code a bit.

File: crates/resolver/src/hosts.rs (hunk 3 above)
Comment (mokeyish) [reply crates/resolver/src/hosts.rs]: This is mainly for unit testing.

image

File: crates/resolver/src/hosts.rs (hunk 4 above)
Comment (mokeyish) [reply crates/resolver/src/hosts.rs]: I don't quite understand this, but I've adjusted it a bit. Please take a look at it again.

File: crates/resolver/src/hosts.rs (hunk 3 above)
Comment (djc) [reply crates/resolver/src/hosts.rs]: In that case the sort should move into the test instead of keeping it in the library.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks!

**Pull request merged**